nexus/src/app/instance.rs

```diff
@@ -1010,6 +1011,23 @@ impl super::Nexus {
     ) -> Result<(), Error> {
         opctx.authorize(authz::Action::Modify, authz_instance).await?;
 
+        // Check that the hostname is valid.
+        //
+        // TODO-cleanup: This can be removed when we are confident that no
+        // instances exist prior to the addition of strict hostname validation
+        // in the API.
+        if db_instance.hostname.parse::<Hostname>().is_err() {
```

Comment:
@smklein This is the shorter-term fix we discussed offline. This is code-path is common to instance start and instance migrate, though because udpates bounce all instances, the former is the relevant one.

It would be nice to beef up the rest of the code (mostly around `nexus_db_model::Instance`) to require a validated hostname everywhere. I'll make an issue if we're happy landing this as-is, and we can tackle that once we're confident no invalid hostnames still exist.

Reply:
Rad, as long as this is visible to customers, this seems like a very appropriate fix to me.

One other option would be validation early in the "instance start" saga -- that's where this `db_instance` comes from anyway, I think, since it's a parameter to that saga -- but this works too, and is about halfway through the saga.

Reply:
Yeah, either would work. I wanted to make sure I didn't miss any code path that calls this, though I don't see any. I'm running a test now to exercise this, manually dorking with the instance record in the DB.